runtime/io.c

```diff
@@ -123,24 +193,17 @@ CAMLexport struct channel * caml_open_descriptor_out(int fd)
   return channel;
 }
 
-static void unlink_channel(struct channel *channel)
-{
-  if (channel->prev == NULL) {
-    CAMLassert (channel == caml_all_opened_channels);
-    caml_all_opened_channels = caml_all_opened_channels->next;
-    if (caml_all_opened_channels != NULL)
-      caml_all_opened_channels->prev = NULL;
-  } else {
-    channel->prev->next = channel->next;
-    if (channel->next != NULL) channel->next->prev = channel->prev;
-  }
-}
-
 CAMLexport void caml_close_channel(struct channel *channel)
 {
   close(channel->fd);
-  if (caml_channel_mutex_free != NULL) (*caml_channel_mutex_free)(channel);
+
   unlink_channel(channel);
+    if (atomic_load_acq(&channel->refcount) > 0) {
+    /* [caml_ml_out_channels_list] may have a reference to this channel. */
+    link_channel (channel);
+    return;
+  }
```

Comment:
I don't understand the logic of this code, and it makes me nervous.  Yes, there can be a race between `caml_close_channel` and `caml_ml_out_channels_list`.  How does this code resolve it?

Reply:
I'm trying to page the details back in. The change is intended to prevent a crash when `flush_all` is executed in parallel with channel finalisation: https://github.com/ocaml-multicore/ocaml-multicore/pull/612/commits/43501299f0c997d4ab1fc557c1745d2f2b9055bb.

The code does look fishy. We should come back to this after the merge.